core/src/impl/Kokkos_HostSharedPtr.hpp

```diff
@@ -0,0 +1,197 @@
+      cleanup();
+      m_element_ptr       = other.m_element_ptr;
+      other.m_element_ptr = nullptr;
+      m_control           = other.m_control;
+      other.m_control     = nullptr;
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION MaybeReferenceCountedPtr& operator=(
+      const MaybeReferenceCountedPtr& other) noexcept {
+    if (&other != this) {
+      cleanup();
+      m_element_ptr = other.m_element_ptr;
+      m_control     = other.m_control;
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+      if (is_reference_counted())
+        Kokkos::atomic_add(&(m_control->m_counter), 1);
+#endif
+    }
+    return *this;
+  }
+
+  KOKKOS_FUNCTION ~MaybeReferenceCountedPtr() { cleanup(); }
+
+  KOKKOS_FUNCTION T* get() const noexcept { return m_element_ptr; }
+  KOKKOS_FUNCTION T& operator*() const noexcept {
+    KOKKOS_EXPECTS(bool(*this));
+    return *get();
+  }
+  KOKKOS_FUNCTION T* operator->() const noexcept { return get(); }
+
+  // checks if the stored pointer is not null
+  KOKKOS_FUNCTION explicit operator bool() const noexcept {
+    return get() != nullptr;
+  }
+
+  // checks whether the MaybeReferenceCountedPtr does reference counting
+  // which implies managing the lifetime of the object
+  KOKKOS_FUNCTION bool is_reference_counted() const noexcept {
+    return m_control != nullptr;
+  }
+
+ private:
+  KOKKOS_FUNCTION void cleanup() noexcept {
+#ifdef KOKKOS_ACTIVE_EXECUTION_MEMORY_SPACE_HOST
+    // If m_counter is set, then this instance is responsible for managing the
+    // objects pointed to by m_counter and m_element_ptr.
+    if (is_reference_counted()) {
+      int const count = Kokkos::atomic_fetch_sub(&(m_control->m_counter), 1);
+      if (count == 1) {
+        (m_control->m_deleter)(m_element_ptr);
+        m_element_ptr = nullptr;
+        delete m_control;
+        m_control = nullptr;
+      }
+    }
+#endif
+  }
+
+  struct Control {
+    std::function<void(T*)> m_deleter;
+    int m_counter;
+  };
+
+  T* m_element_ptr;
+
+ protected:
+  Control* m_control;
+};
+
+template <class T>
+class HostSharedPtr : public MaybeReferenceCountedPtr<T> {
+ public:
+  // Objects that are default-constructed or initialized with an (explicit)
+  // nullptr are not considered reference-counted.
+  HostSharedPtr() noexcept : MaybeReferenceCountedPtr<T>(nullptr) {}
+  HostSharedPtr(std::nullptr_t) noexcept : HostSharedPtr() {}
+
+  explicit HostSharedPtr(T* element_ptr)
+      : MaybeReferenceCountedPtr<T>(element_ptr, [](T* const t) { delete t; }) {
+  }
+
+  template <class Deleter>
+  explicit HostSharedPtr(
+      T* element_ptr, Deleter deleter = [](T* const t) { delete t; })
+      : MaybeReferenceCountedPtr<T>(element_ptr, std::move(deleter)) {}
+
+  int use_count() const noexcept {
+    return this->m_control ? this->m_control->m_counter : 0;
+  }
+};
+
+template <class T>
+class UnmanagedPtr : public MaybeReferenceCountedPtr<T> {
+ public:
+  explicit UnmanagedPtr(T* element_ptr = nullptr) noexcept
```

Comment:
The default constructor should be separate and non-`explicit`.

Reply:
OK, if that helps moving forward. This doesn't make any real difference, does it?